### PR TITLE
[ELY-1115] null principal handling in ServerAuthenticationContext

### PR DESCRIPTION
**Pull request by honza889**
Fix of RegexValidatingPrincipalTransformerTestCase broken by [ELY-1115].

Null principal should be considered wrong username passed by user, as discussed.

Replaces https://github.com/wildfly-security/wildfly-elytron/pull/798 and https://github.com/wildfly-security-incubator/wildfly-core/pull/156

### CODE REVIEW

File: src/main/java/org/wildfly/security/auth/server/ServerAuthenticationContext.java [hunk 1]
```diff
@@ -1010,12 +1004,24 @@ NameAssignedState assignName(final SecurityIdentity capturedIdentity, final Mech
     NameAssignedState assignName(final SecurityIdentity capturedIdentity, final MechanismConfiguration mechanismConfiguration, final MechanismRealmConfiguration mechanismRealmConfiguration, Principal originalPrincipal, final Evidence evidence, final IdentityCredentials privateCredentials, final IdentityCredentials publicCredentials, final boolean exclusive) throws RealmUnavailableException {
         final SecurityDomain domain = capturedIdentity.getSecurityDomain();
         final Principal preRealmPrincipal = rewriteAll(originalPrincipal, mechanismRealmConfiguration.getPreRealmRewriter(), mechanismConfiguration.getPreRealmRewriter(), domain.getPreRealmRewriter());
+        if (preRealmPrincipal == null) {
+            log.tracef("Unable to rewrite principal [%s] by pre-realm rewritters", originalPrincipal);
+            return new NameAssignedState(capturedIdentity, null, RealmIdentity.NON_EXISTENT, preRealmPrincipal, mechanismConfiguration, mechanismRealmConfiguration, privateCredentials, publicCredentials);
```
Comment (dmlloyd) [inline src/main/java/org/wildfly/security/auth/server/ServerAuthenticationContext.java]: I'm not sure we should enter name-assigned state in this case though.

Comment (dmlloyd) [inline src/main/java/org/wildfly/security/auth/server/ServerAuthenticationContext.java]: If the rewrite fails at this stage, it means the user name is invalid and there can be no assigned name.  So the resultant stage should be a subclass of UnassignedState - possibly an InvalidNameState even, if we need special behavior in this case (for example we would need authorization to fail).

Comment (honza889) [inline src/main/java/org/wildfly/security/auth/server/ServerAuthenticationContext.java]: Would not be FAILED state sufficient?

**Comment by honza889** [comment]
Updated to use new  InvalidNameState as requested by @dmlloyd. (wildfly+wildlfy-core green)

**Pull request merged**